File: problems/040-dual-quota-output-knapsack/solutions/python/main.py
import sys

t = list(map(int, sys.stdin.buffer.read().split()))
it = iter(t)
N, B, I = next(it), next(it), next(it)
dp = [[0] * (B + 1) for _ in range(I + 1)]
for _ in range(N):
    b, e, v = next(it), next(it), next(it)
    for x in range(I, e - 1, -1):
        row, old = dp[x], dp[x - e]
        for y in range(B, b - 1, -1):
            row[y] = max(row[y], old[y - b] + v)
print(dp[I][B])

File: problems/040-dual-quota-output-knapsack/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let n: usize = t.next().unwrap().parse().unwrap();
    let bmax: usize = t.next().unwrap().parse().unwrap();
    let imax: usize = t.next().unwrap().parse().unwrap();
    let mut dp = vec![vec![0u64; bmax + 1]; imax + 1];
    for _ in 0..n {
        let b: usize = t.next().unwrap().parse().unwrap();
        let e: usize = t.next().unwrap().parse().unwrap();
        let v: u64 = t.next().unwrap().parse().unwrap();
        for x in (e..=imax).rev() {
            for y in (b..=bmax).rev() {
                dp[x][y] = dp[x][y].max(dp[x - e][y - b] + v);
            }
        }
    }
    println!("{}", dp[imax][bmax]);
}
